### DISCUSSION TITLE
Get Vanara to show properties of COM objects in Visual Studio debugger?

### DISCUSSION THREAD

**Discussion by aybe** [dahall/Vanara#265] [Q&A]
Using the IMAPI package, when you debug the program, you can't see any of the properties for a COM object in the debugger:

<img width="440" alt="ACDSeeQVUltimate15_O37ijUw3hQ" src="https://user-images.githubusercontent.com/1537591/147769473-cbb7fb81-66a7-4f6d-ad8f-b917ab4d93cb.png">

On the other hand, when you manually add a COM reference using the Visual Studio wizard, during debugging you can see the properties of COM objects:

<img width="394" alt="devenv_2x7Tq4cnuq" src="https://user-images.githubusercontent.com/1537591/147769960-3055ba31-bb74-4e97-a62f-5ba9b2038721.png">

<img width="445" alt="ACDSeeQVUltimate15_D4fEzSpjp8" src="https://user-images.githubusercontent.com/1537591/147769536-9bfbc821-48a9-4664-9f04-09d9b4f843fb.png">

But then of course, when going this way one loses all of the fantastic plumbing and helpers provided by Vanara...

Any ideas on how to have the best of both worlds, i.e. get Vanara to show properties of COM objects in debugger?

Thank you!

**Comment by dahall** [comment]
The answer, I believe, is in the first graphic: "To inspect the native object, enable native code debugging." The objects exposed by Vanara are native COM objects, with no Runtime Callable Wrappers. The objects exposed when adding a COM reference are RCWs and have special handling within .NET. See [here](https://www.technipages.com/visual-studio-native-code-debugging) for instructions.

**Comment by aybe** [comment]
Hello and happy new year!

I've tried your suggestions and I get the following debugging experience:

<img width="468" alt="ACDSeeQVUltimate15_hQK3k42zrv" src="https://user-images.githubusercontent.com/1537591/147885223-45a006b8-114d-4b81-8149-9d8359efedb1.png">

So it shows something but none of the actual properties, I suppose it's expected as these are methods on native side.

As I understand you, it's these RCWs that make the properties available during debugging, correct ?

Which leads me to ask you, could a custom build of Vanara leverage these RCWs ? or anything else

Thank you.
